platform-sdk/docs/proposals/state-lifecycle/proposal.md

```diff
@@ -0,0 +1,86 @@
+
+### Requirements
+
+- Add an interface representing state lifecycle management and its implementation.
+- Update the current State API to be coherent with this State Lifecycle API.
+- Implement this set of APIs and migrate the existing code to use it.
+
+## Design & Architecture
+
+### Java classes
+
+- `com.swirlds.state.State` has the following lifecycle-related methods:
+  - `State copy()`: creates a copy of the state, which is mutable. The source of the copy becomes immutable.
+  - `void computeHash()`: computes the hash of the immutable state.
+  - `void createSnapshot(Path targetPath)`: creates a snapshot of the state at the specified path.
+  - `State loadSnapshot(Path sourcePath)`: loads the snapshot from the specified path.
+
+These methods correspond to the state lifecycle as follows:
+
+[![State lifecycle](state-lifecycle.svg)](state-lifecycle.svg)
+
+Note that the `loadSnapshot` method doesn't really belong to `com.swirlds.state.State`, as it's a method of the class, and in some cases, we may need to load the snapshot before we have an instance of the state. 
+This method should be moved to a separate class that will be responsible for managing the state lifecycle.
+
+- `SwirldStateManager` needs refactoring. The current implementation has two sets of responsibilities:
+  - Keeping track of the references to the latest immutable state and the latest mutable state. This includes creating a copy of the state and updating the references.
+  - Handling Platform-related events:
+    - Handling consensus rounds.
+    - Sealing consensus rounds.
+    - Keeping the `freeze time` parameter up-to-date.
+    - Providing state for signing.
+    - Initializing the state by extracting it from the signed state.
+
+These two sets of responsibilities should be separated. The `swirlds-state-impl` should not have any details related to the platform. However, it should provide all necessary levers for the platform to interact with the state and its lifecycle.
+
+Therefore, `swirlds-state-impl` needs a `StateManager` class that will have the following responsibilities:
+- Keeping track of the references to the latest immutable state and the latest mutable state.
+- Tracking states that existed before the latest immutable state. These should be available for lookup by their round number.
+- Evicting old states that are no longer needed.
+- Loading snapshots from the disk.
+
+Additionally, there should be a configuration parameter that limits the number of states kept in memory. When the limit is reached, the oldest state should be evicted.
+
+The interface should look as follows:
+
+```java
+public interface StateManager {
+    State getLatestImmutableState();
+    
+    State getLatestMutableState();
+    
+    State getState(long round);
+    
+    void evictState(long round);
```

Comment:
This one is questionable... In the current implementation, states are released using ref counting mechanism. Will it be different in `swirlds-state-impl`?

Reply:
Yes, we want to get rid of the reference counting all together. The state module will retain a state in memory until it is told to release it, or possibly until it reached the configured maximum age, although i'm not sure about this second part.

Reply:
I'll explicitly mention that `evictState` is there as a replacement for the reference counting mechanism